test_tools/onnx_pytorch/common/parser.py

```diff
@@ -0,0 +1,234 @@
+
+        value = evt.Value()
+        if evt.ValueType() == OutcomeValue.literal:
+            value = self.getString(value)
+        elif evt.ValueType() == OutcomeValue.numeric:
+            value = self.cast(value, NumericOutcome).Value()
+
+        index = None
+        if evt.ActionTaken() is True:
+            index = evt.Index()
+            if evt.IndexType() == OutcomeValue.literal:
+                index = self.getString(index)
+            elif evt.IndexType() == OutcomeValue.numeric:
+                index = self.cast(index, NumericIndex).Index()
+
+        return value, index
+
+    def apply_reward_fn(self, rewards):
+        reward_fn_type = self.checkpoint_info.RewardFunctionType()
+        if len(rewards) == 0:
+            return self.checkpoint_info.DefaultReward()
+        elif reward_fn_type == RewardFunctionType.Earliest:
+            return rewards[0]
+        elif reward_fn_type == RewardFunctionType.Average:
+            return sum(rewards)/len(rewards)
+        elif reward_fn_type == RewardFunctionType.Median:
+            from statistics import median
+            return median(rewards)
+        elif reward_fn_type == RewardFunctionType.Sum:
+            return sum(rewards)
+        elif reward_fn_type == RewardFunctionType.Min:
+            return min(rewards)
+        elif reward_fn_type == RewardFunctionType.Max:
+            return max(rewards)
+        else:
+            raise Exception("Unknown reward function type")
+        
+
+    # reads all events associated with the next event_id
+    def read_event_series(self, timestamp=None):
+        for msg in self.reader.messages():
+            # only support CB for now
+            label = None
+            rewards = []
+            current_id = None
+            current_payload = None
+
+            # One thing to note, the chosen action is either encoded in the ActionIds array
+            # in the CB payload as the first element (if DeferredAction == False)
+            # or its in the outcomes payload as the ActionTaken if DeferredAction == True
+            for i in range(msg.EventsLength()):
+                joined_event = msg.Events(i).EventAsNumpy()
+                evt = Event.GetRootAsEvent(joined_event, 0)
+                m = evt.Meta()
+                event_payload = evt.PayloadAsNumpy()
+            
+                if m.Encoding() == EventEncoding.Zstd:
+                    event_payload = zstd.decompress(event_payload)
+
+                if m.PayloadType() == PayloadType.CB:
+                    # ew gross
+                    if label is not None:
+                        cost = -1*self.apply_reward_fn(rewards)
+                        yield current_payload, label, cost
+                    current_id = m.Id()
+                    current_payload, label = self.parse_cb(event_payload)
+                elif m.PayloadType() == PayloadType.Outcome:
+                    tmpreward, tmplabel = self.parse_outcome(event_payload)
```

Comment:
Did you remove the label - 1? Is it not needed now?

Reply:
The -1 is in CbDictParser. I'm not sure where the best place to put it was. For now, I put it at the closest point to the caller, but the closest point to the source (in parse_outcome) is another possibility. Any preference?

Reply:
Ah, actually there are 2 points where we can get the index. `parse_outcome` for delayed actions and `parse_cb` for non-delayed actions. So I put the logic in CbDictParser to have a single location to do the math